### ISSUE TITLE
Application Flow

### ISSUE THREAD

**Issue by SimonLab** [dwyl/github-backup#11]
## Github App send events
The Github app will send the following events to the server:
- new installation on repository
- new issue
- new comment on issue
- issue's description edited
- comment edited

Questions:
- A Github App can also be installed on an organisation, will this trigger an event for all the repository of the organisation? We need to get more details on how webhooks work with organisations and test it once we have a server running. see https://developer.github.com/webhooks/#events
- Is the "issue edited" and "comment edited" events are the same? I think the events are different: https://developer.github.com/v3/activity/events/types/#issuecommentevent and https://developer.github.com/v3/activity/events/types/#issuesevent

## Server receive events

The server will receive a post request from the Github App for each Github app webhooks events (see above). The main role of the server is to determine the type of the events, this can be done based on the payload (and often the "action" value of the payload). Then the information of the payload is saved in database and files (see https://github.com/dwyl/github-backup/issues/3#issuecomment-365650695). Once the information is saved the server can add more information on the issue to let the Github users know how to access the backup data.

Questions:
- Which information needs to be saved in the database, what is the architecture of the Postgres database?
- What are the information and structure (json?) of the files we want to create?

##  Add "backup" link to issues
Once a backup of the issue has been created the server will update the issue/comment with a link to the "backup" app

Questions:
- How to represent the link to the backup app on Github? Edit the issue/comment and add a link at the bottom? Can we apply a template?
- Should the link be display on the description of the issue and on each comment or only on the issue itself?

## See backup

The users of a repository can access the backup of the issues/comments. The application needs to check the permission of the users and then display the information

Questions:
- Who can access the "backup" app, see https://github.com/dwyl/github-backup/issues/7#issuecomment-365691051 ?
- How to represent the history of the edits? Download json file only or display a list of edit sorted by date?

I try to describe the main steps of the flow, some parts need a lot of more details and we can create new issues for them. I am not sure especially on the last two section: how to add and represent the link to the "backup" app on Github issues and how the "backup" app display the information to the users. Let me konw which steps I have missed!

**Comment by nelsonic** [comment]
@SimonLab / @Cleop  can you please _check/confirm_ if you are able to `edit` this Drawing:
https://drive.google.com/file/d/1EHjLPs6N2P_d49AVRfbYKDhWVXaiZtHg/view
It should open in `draw.io` and requires that you are logged in with your @dwyl.com Google Account.

![github-backup-flow](https://user-images.githubusercontent.com/194400/36377704-f8af4c38-156f-11e8-90f8-18d5656274c5.png)

My _Plan_ is to open an issue for each of these "boxes" and then add the Issue Number to the Diagram.

**Comment by Cleop** [comment]
@nelsonic - yes I can now I've authorised `draw.io` 👍

**Comment by SimonLab** [comment]
@nelsonic works for me too, thanks

**Comment by Cleop** [comment]
Nice diagram @nelsonic - What data is classified as meta and what is classified as rich text? Is rich text literally the comment and meta is all the data around when the comment was made, by whom etc?

**Comment by nelsonic** [comment]
@Cleop precisely. The Comments contain markdown and take up _lots_ of space in the DB,
so we want to store them in files (_or some other "cheap" storage_)
